### ISSUE TITLE
more metadata needed?

### ISSUE THREAD

**Issue by npomfret** [timmolter/XChange#1544]
I've just tried executing a trade on the BTC/LTC market on hitbtc which failed with the message 'Rounding necessary'.  Now it seems that LTC and BTC are represented on the 8 exchange with 8 decimal places, as is the price on the BTC/LTC exchange.  However, the quantity that is tradable on that market is rounded to just 1 decimal place.  Is there a common place in this API (like the .json metadata files) to specify that value?

I was thinking something like:

```
{
  "currency_pairs": {
    "LTC/BTC": {
      "price_scale": 8,
      "quantity_scale": 2
    }
  }

  ...

  "BTC": {
    "scale": 8
  },
  "LTC": {
    "scale": 8
  },
```

**Comment by timmolter** [comment]
The metadata may or may not have specifically that, I don't know for sure as I've not yet used that specific feature. The json could also very well be outdated. You'll have to investigate it yourself unless someone else chimes in here.

**Comment by npomfret** [comment]
Thanks for the reply - I'll give it some thought.

**Comment by molexx** [comment]
I'd be interested to know any progress/thoughts on this, thanks.

**Comment by npomfret** [comment]
I'm quite new to this project, so my ideas might not be quite thought out enough yet... 

It feels like in order to trade on any exchange, each needs some common metadata.  Specifically, so far I've come across:

_currency scale_ - for use when getting your wallet data
_price scale_ - for use when making a limit order
_quantity scale_ - for use when setting the quantity/size of any order (which doesn't seem to be represented anywhere)

And I think these should be present in the exchange metadata for __all__ exchanges.

_fees_ - these vary from exchange to exchange.  Some have taker v's maker fees, some have none at all (coin floor for example).  They are usually expressed as a percentage of trade size.  But some exchange have more complex models, with discounts for example.  Some have funding fees, as well as trade fees. And sometimes fees vary based on what's being traded (as well as the trade size).

I think maybe the fees are a bit trickier to model and perhaps should be removed from the static data al together? 

I would be very useful if all trades returned by `getTradeHistory` would give amounts without fees applied and so the user can apply the fees separately.  So far I've sound some implementations that do as I've described, and some that give amounts net (after fees have been removed from the trade value).  This inconsistency makes it difficult to do accounting. 

In addition, we also may have:

_rate limits_ - to prevent a client spamming the exchange with requests.  I think these are typically easy to model and probably belong in the static data.

I've discovered that each sub-module has a `exchange-name.json` at the root of the class path that contains the static data.  It would have been helpful (for a beginner like me) if this was documented in the `readme.md` file.  I've had to create my own copies of almost every one I've used because they're out of date.  And also, I've discovered that some exchanges provide _some_ of this metadata in the api, and those cases the static data can be overwritten by the dynamic data.

I wonder if the static data, which seems to change quite frequently might be best kept in a separate repository?  That way they could be updated frequently by this community and users would have to wait for a release.

Just some thoughts... feel free to ignore me :)

**Comment by timmolter** [comment]
Thanks, @npomfret . I think I may put some of what you wrote in the Wiki. I'll comment more on it soon...

**Comment by molexx** [comment]
Yes a separate repo with the json files in sounds like a good idea. As they are loaded from the classpath it shouldn't require any changes other than moving the files around in git and adding a mvn module.
I like the current method (documented in the wiki) of merging files with data pulled from online APIs. It would be good for data to be removed from the .json if an API source is implemented.

Maker/taker fees could be done by splitting `CurrencyPairMetaData.tradingFee` into two e.g. `makerTradingFee` and `takerTradingFee`. The change could be made non-breaking by having `tradingFee` pick one to return but I'd probably be in favour of a breaking change in order to have accurate data.
Tiered fee levels sound complex to model but yes it does seem to be common.

Deposit/withdraw fees are on the currency rather than the currencypair and are usually a fixed amount (e.g. 0.0001 BTC). Maybe for some currencies there's a percentage-based fee too but I can't remember seeing one other than for fiat... which gets quite complex with different payment providers (OKPay etc.) and so might be easier to leave out at least initially. 
If fixed amounts are all we need for now it should be straightforward to add something like `depositFee` and `withdrawlFee` to `CurrencyMetaData.class`.
Most (all?) currencies have a 'minimum network fee' for sending funds, some exchanges just quote that as the withdrawl fee (so don't actually charge any extra for themselves) and some add a bit on top. Maybe we'd like to model those two separately?

Public and private rate limits are I think already modelled fully in `ExchangeMetaData.class` but of course it's a matter of filling the .json data for every exchange.

I don't know what any of the 'scaling' values mean or are/should be used for, please could I have a quick description? :-)

**Comment by npomfret** [comment]
WRT the fees modelling...

I agree that funding (deposit and withdrawal fees) are probably best ignore by this project, at least for now.

And I think to model the fees better then separating taker from maker would be a good step forward.  In addition I think we need per currency-pair fees as some exchanges have different fees in different markets.  

And... the fees are sometimes issued in the base currency, and sometimes the counter currency.  Sometimes the user can choose which (and it's a port of your account settings).

@molexx the 'scaling' parameters are really important.  If you don't scale your values properly the exchanges will reject your api calls.  The error messages will say something like "needs rounding" or "price too accurate".  Basically if you send a value at the wrong precision the exchange will either round it or reject it - I think its common practice to reject it.

So we have

 * currency-scale - for use when depositing and withdrawing currency.  Possibly you'd also need use this to calculate your fees accurately
 * price-scale - for use when trading
 * amount-scale - doesn't exist in this project yet.  Also for use when trading to set the quantity on your order.  Maybe tick-size would be a better name. In some markets the trade size / quantity needs to be divisible by 100 or even 1000.

**Comment by timmolter** [comment]
@npomfret @molexx Thanks for your thoughts on the matter. 

>  It would have been helpful (for a beginner like me) if this was documented in the readme.md file.

https://github.com/timmolter/XChange/wiki/Design-Notes

See section on metadata. I should probably add an easy-to-see link in the README.

> I wonder if the static data, which seems to change quite frequently might be best kept in a separate repository?

Well, the `develop` branch is more or less that. Plus the code is also changing so much chasing changes on exchanges' API. Every time I do a release there are hotfixes and additions literally the next day. :) It's tricky because there is no real way to run integration tests on all the code because we're dealing with live exchanges and API keys to private accounts.

> I've had to create my own copies of almost every one I've used because they're out of date.

Please contribute your changes, as that's the only way we can hope to keep things updated. 

> And also, I've discovered that some exchanges provide some of this metadata in the api, and those cases the static data can be overwritten by the dynamic data.
> I like the current method (documented in the wiki) of merging files with data pulled from online APIs. It would be good for data to be removed from the .json if an API source is implemented.

In all cases I've seen the dynamic data is used to supplement the static data. Where the dynamic data can be accessed it's considered to me more accurate than the static data and would override it, yes.

> fees
> but I'd probably be in favour of a breaking change in order to have accurate data.

Maker, taker and tiered fees would be a great addition to the metadata as it would help inform bots/algos/people before placing an order. Additionally, I think in many cases the fees charged are given with the user trades, which can be used for book keeping. A breaking change is fine.

> rate limits 

Those are there already.

> we need per currency-pair fees as some exchanges have different fees in different markets.

Those are there already. mostly I suppose.

> And... the fees are sometimes issued in the base currency, and sometimes the counter currency. Sometimes the user can choose which (and it's a port of your account settings).

yes, complicated. But could be done.

> amount-scale - doesn't exist in this project yet.

Definitely could be added.

**Comment by timmolter** [comment]
referencing this related issue: https://github.com/timmolter/XChange/issues/1617

**Comment by jnorthrup** [comment]
I consider #2507 to fall into the category of "general Metadata Improvements"
